arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/Bracket.kt

```diff
@@ -231,31 +206,35 @@ public suspend inline fun <A, B> bracketCase(
   use: suspend (A) -> B,
   crossinline release: suspend (A, ExitCase) -> Unit
 ): B {
-  val acquired = withContext(NonCancellable) {
-    acquire()
-  }
-
-  val res = try {
-    use(acquired)
-  } catch (e: CancellationException) {
-    runReleaseAndRethrow(e) { release(acquired, ExitCase.Cancelled(e)) }
-  } catch (t: Throwable) {
-    runReleaseAndRethrow(t.nonFatalOrThrow()) { release(acquired, ExitCase.Failure(t.nonFatalOrThrow())) }
-  }
-
-  withContext(NonCancellable) { release(acquired, ExitCase.Completed) }
-
-  return res
+  val acquired = withContext(NonCancellable) { acquire() }
+  return guaranteeCase({ use(acquired) }) { release(acquired, it) }
 }
 
 @PublishedApi
-internal suspend inline fun runReleaseAndRethrow(original: Throwable, crossinline f: suspend () -> Unit): Nothing {
+internal suspend inline fun runReleaseAndRethrow(original: Throwable?, crossinline f: suspend () -> Unit) {
   try {
     withContext(NonCancellable) {
       f()
     }
   } catch (e: Throwable) {
-    original.addSuppressed(e.nonFatalOrThrow())
+    original?.addSuppressed(e.nonFatalOrThrow()) ?: throw e
+  }
+  original?.let { throw it }
+}
+
+@PublishedApi
+internal inline fun <R> finalizeCase(block: () -> R, finalizer: (ExitCase) -> Unit): R {
```

Comment:
Should we make this public?
Also annoyingly it duplicates the bytecode of `finalizer`. I think? this can be avoided

Reply:
it seems fairly delicate with respect to exception/suppression. The only benefit over `guaranteeCase` is it's non-suspending.

Reply:
Very true. It's however exactly the right method needed to use `allocate` correctly, or to use any `release`-like method that deals with suppression and non-cancellability, but it is very niche, I'll admit

Reply:
The original purpose of `allocate` was to break up a resource lifecycle for integrations (such as test listeners) which wouldn't be able to use `finalizeCase` for the same reason it can't just use `.use`. Gut says to keep it internal for now, and expose it should a viable usecase come about. Either that, or give it a suitable opt-in annotation.